docs/architecture/adr-009-revamp-testkit.md

```diff
@@ -0,0 +1,279 @@
+# ADR 009: Revamp IBC integration test framework
+
+## Changelog
+
+- 04-03-2024: Initial draft
+
+## Context
+
+The current framework in the IBC testkit uses
+[existing types and injects state or dependency
+manually](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/tests/core/ics02_client/update_client.rs#L574-L578).
+Sometimes, it uses
+[semantically wrong data as mock data](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/src/testapp/ibc/core/types.rs#L320).
+Because of this, tests with customizable steps and fixed fixtures became ad-hoc
+and unmaintainable.
+
+To overcome this, we need to improve our test framework that allows:
+
+- testing different implementations (traits).
+- succinct tests (useful `util` methods).
+- improving test coverage (Merkle proof generation).
+- integration tests exercising the IBC workflow (relayer-like interface)
+
+## Decision
+
+The main goal of this proposal is to create a test framework that is modular and
+closer to the real blockchain environment. This should also make the existing
+tests succinct and readable. Instead of bootstrapping the mock data, we should
+use valid steps to generate it - so that we know the exact steps to reach a
+state to reproduce in a real environment.
+
+To achieve this, we have broken down the proposal into sub-proposals.
+
+### Adopt a Merkle store for the test framework
+
+The current framework uses `HashMap` and `HashSet` to store data. This works for
+many test scenarios, but it fails to test proof-sensitive scenarios. Because of
+this, we don't have any connection, channel handshake or packet relay tests for
+Tendermint light client.
+
+We generalize
+[`MockContext`](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/src/testapp/ibc/core/types.rs#L103)
+to use a Merkle store which is used for IBC Context's store. For concrete or
+default implementation, we can use the IAVL Merkle store implementation from
+`informalsystems/basecoin-rs`.
+
+### Modularize the host environment
+
+Currently, we are using `Mock` and `SyntheticTendermint`
+[variants](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/src/hosts/block.rs#L33-L36)
+as the host environments. To manage these two different environments, we also
+introduced
+[`HostBlocks`](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/src/hosts/block.rs#L72-75)
+for corresponding host variants.
+
+This creates friction if we have to add a new host variant. It creates the same
+problem as why we have `ibc-derive` crate for `ClientState` and
```

Comment:
```suggestion
problem that the `ibc-derive` crate was designed to solve for `ClientState` and `ConsensusState` types.
```

Reply:
Additionally, can we summarize the actual problem here, e.g.: "It creates the same problem that the `ibc-derive` crate was designed to solve for `ClientState` and `ConsensusState` types, namely that: `[problem summary]`"

Reply:
d86ddadb